Python/chapter01/1.3 - URLify/miguel_1.3_sol.py

```diff
@@ -0,0 +1,96 @@
+"""
+Python version 3.7.0
+1.3 - URLify
+Write a method to replace all spaces in a string with "%20".
+The following only applies to C/C++: "You may assume that the string has sufficient
+space at the end to hold the additional characters, and that you are given the "true" length of the string"
+(Note: If implementing in Java, please use a character array so that you can perform this operation in place)
+Since we are using python, no need to use true length.  I will have two functions, one with
+true_length and the other without.
+"""
+import unittest
+
+
+def urlify(s: str, true_length: int) -> str:
+    """
+    Given a string and it's "true" length, this function will return a new string
+    that replaces all of the spaces of the input string with '%20'.
+    Precondition(s):
+    - length of s <= true_length
+    :param s: the original string to 'urlify'
+    :param true_length: since s may have additional characters, we focus on true_length instead of actual s length
+    :return: string with each space from s replaced with '%20'
+    """
+    # Below link goes over string concat efficiency in python (I will use method 4
+    # https://waymoot.org/home/python_string/
+    # each space will make us replace it (with '%', then add 2 characters ('2', '0')
+    result = ['\x00'] * len(s)*2  # extra buffer space
```

Comment:
for this challenge to be successful, you cannot use extra buffer space:

```suggestion
    result = ['\x00'] * true_length
```

the contract in C/C++ is that you **must** tell library functions the truth: if the function says you need to specify the length of the buffer in a way that the final result fits, you should trust that. any crashes that happen because you were lied to are the caller's fault, not yours.

in any case, if you pass a string with a single space, your version would crash since it needs 3, not 2 slots.

Reply:
For the final result to fit, shouldn't the size be true_length + 2*space_count?

Going to add an extra test case for single space.

Reply:
In the example input/output:
input:  "Mr John Smith    ", 13
output:  "Mr%20John%20Smith"

let's assume that true_length tells us what the output length should be.

Based on the example, that would mean that the output string must be of size 13.
The length of the output is 17, we have a contradiction.

So true_length is not the length of what the output should be (according to the author, at least)

Or maybe it could be that the author made a typo.